hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java

```diff
@@ -50,8 +50,25 @@
    * @param schema Schema used for record
    * @return new combined/merged value to be written back to storage. EMPTY to skip writing this record.
    */
+  @Deprecated
   Option<IndexedRecord> combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema) throws IOException;
 
+  /**
+   * This methods lets you write custom merging/combining logic to produce new values as a function of current value on
+   * storage and whats contained in this object.
+   * <p>
+   * eg: 1) You are updating counters, you may want to add counts to currentValue and write back updated counts 2) You
+   * may be reading DB redo logs, and merge them with current image for a database row on storage
+   *
+   * @param currentValue Current value in storage, to merge/combine this payload with
+   * @param schema Schema used for record
+   * @param props Payload related properties. For example pass the ordering field(s) name to extract from value in storage.
+   * @return new combined/merged value to be written back to storage. EMPTY to skip writing this record.
+   */
+  default Option<IndexedRecord> combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema, Map<String, String> props) throws IOException {
```

Comment:
Can you describe any more use-cases apart from the ordering field for the Map<> introduced ?

Reply:
Map was a more generic structure to send properties. For now, its only the ordering field that is required here. For your second question, not sure if I am understanding it right.  Are you suggesting, we send in the ordering field val itself instead of the field name ?

Reply:
Let me explain. The current interface `combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema)` doesn't allow the user to pass the ordering field names. You're trying to introduce this and have a Map<> to allow the user to do this. (The description link in the JIRA ticket doesn't work so I needed more context). 
Now, since the user can have their their own implementation of `combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema)`, they can have the ordering field name in their own class and maintain it themselves, thus not requiring the need to explicitly pass these names through a method. 
I see that the intention of this PR to support this functionality specifically for the `OverwriteWithLatestAvroPayload`. For preCombine, will we just continue to support natural ordering which may not be the same as for combineAndGetUpdateValue, correct ?

Reply:
Thanks for the clarification @n3nash . I am not able to open the link either. It could be because the initial reporter had different jira ids or something else. But the comments section has good context on what this ticket is about. Also, this has come across multiple times in Slack channels from different users. Based on that I can summarize as follows.

 Users are looking to consider records in disk when using  `OverwriteWithLatestAvro` payload class. In `preCombine` step - we pick the latest value for every key from a batch of input data based on some ordering field. At this step we dont know what is in storage yet. When we are ready to write the records, we iterate the records in storage, for each record determine if there is a matching entry in the input batch. If so, we invoke `combineAndGetUpdateValue` and pass in the record in disk as the `currentValue` param. In this step specifically, `OverwriteWithLatestAvro` can possibly overwrite an already stored record with an older update since we werent comparing the ordering val of new data with data in disk. This whole PR is to provide that capability instead of asking Users to write their own implementation to do this. w.r.t `preCombine` and `combineAndGetUpdateValue` both will be using the same ordering field(s) to determine the latest. 

In order to not disrupt other payload classes, I deprecated the existing `combineAndGetUpdateValue`, extended it and provided a default implementation that would ignore the Map field and internally do the old way. `OverwriteWithLatestAvroPayload` alone will override this default functionality to achieve the above purpose.

Hope that helps!

Reply:
@bhasudha thanks for explaining, although my metapoint is still not resolved. If you look at the log scanner -> https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java#L114, the preCombine method is used to combine records written to the log (over multiple batches). Ideally, according to your use-case, since during logging we don't know what is the order of updates, we simply write to disk and leave the resolution for a later time. In this scenario, since `combineAndGetUpdateValue` is not called, how will `preCombine` honor the `orderingVal` ?

Reply:
@n3nash  preCombine is using the same orderingVal as would combineAndGetUpdateValue. And in the HoodieMergedLogRedordScanner class we are trying to resolve records that are logged to the system already and use orderingVal to identify the most recent one. I dont understand why would we need to use combineAndGetUpdateValue here.

Reply:
Just my two cents:
The interface changes looks a bit of overkill for this type of new feature.

If we are still relying on avro schema, we may leverage `schema.getProp(orderingFieldKey)` to get ordering field for this record.

Reply:
@advancedxy I guess we would nt know which is the orderingFiledKey when calling this method. That is why we would need to send in that info using another param, so we can get fetch that from the schema later.